### PR TITLE
Switch the the UI thread to get the VS version

### PR DESCRIPTION
**Pull request by joelverhagen**
Fix https://github.com/NuGet/Home/issues/3419

@kzu, I took a slightly different tack than you took in https://github.com/NuGet/NuGet.Client/pull/871. Does this also look reasonable?

/cc @kzu @emgarten @alpaix @rohit21agrawal

### CODE REVIEW

**Comment by alpaix** [comment]
@joelverhagen Why is it necessary setting `UserAgent` from within `VsSolutionManager`? Can it be set in `NuGetPackage.Initialize` instead?

**Comment by joelverhagen** [comment]
@alpaix, good idea! I considered this but wanted to minimize my changes (reducing risk) because I wanted to push for this in 3.5.

**Comment by kzu** [comment]
As long as setting that version isn't required for interacting with the nuget services, it should be fine. Otherwise, you'd be forcing consumers of your APIs to ensure the package is initialized before using them, which would no be intuitive or friendly at all

**Comment by alpaix** [comment]
MEF import via `SVsServiceProvider` guarantees the package is initialized ahead of any MEF export got instantiated. Hence package's `Initialize` method is the right and the safe place to relocate this statement to.

File: src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 1]
```diff
@@ -102,8 +102,22 @@ public VSSolutionManager()
                 ErrorHandler.ThrowOnFailure(hr);
             }
 
-            UserAgent.SetUserAgentString(
-                new UserAgentStringBuilder().WithVisualStudioSKU(VSVersionHelper.GetFullVsVersionString()));
+            // Allow this constructor to be invoked from either the UI or a worker thread
+            if (ThreadHelper.CheckAccess())
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: This shouldn't be needed. From what I have been told starting a JTF while already on the UI thread is a noop and very fast.

**Review by emgarten — DISMISSED** [review]

File: src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Got the same feedback from Yishai. I was being overly careful it seems 😄. Will fix.

Comment (joelverhagen) [reply src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Fixed.

**Comment by joelverhagen** [comment]
> Hence package's Initialize method is the right and the safe place to relocate this statement to.

I tested this. Not the case in dev14 at least. The first that is called from this line:

```
components.DefaultExportProvider.GetExportedValue<object>("NuGet.PackageManagement.ISolutionManager");
```

... is the constructor of `VSSolutionManager`.

**Comment by kzu** [comment]
One way I've solved this in the past is having the component ctor get the IVsShell from the SVsServiceProvider and call IsPackageLoaded/LoadPackage to ensure its providing package is initialized

**Comment by alpaix** [comment]
The key thing was to acquire component model instance using `SVsServiceProvider`.

File: src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 2]
```diff
@@ -102,8 +102,14 @@ public VSSolutionManager()
                 ErrorHandler.ThrowOnFailure(hr);
             }
 
-            UserAgent.SetUserAgentString(
-                new UserAgentStringBuilder().WithVisualStudioSKU(VSVersionHelper.GetFullVsVersionString()));
+            // Allow this constructor to be invoked from either the UI or a worker thread
+            ThreadHelper.JoinableTaskFactory.Run(async delegate
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:107]: So in order to support Xamarin we're about to sacrifice perf in the rest 99% of cases here?

**Comment by kzu** [comment]
As multiple other reports on the linked bug show, this is far from a
Xamarin issue (which we have dealt with, btw, but everyone else is still
broken because if this)

**Comment by joelverhagen** [comment]
@kzu, could you provide a representative example of what Xamarin was doing to import NuGet prior to your fix? I would like to manually verify that my change fixes the problem. I have already verified this with your minimal repro snippet and with Service Fabric.

**Comment by emgarten** [comment]
:shipit:

**Review by dtivel — APPROVED** [review]

**Pull request merged**